File: doc-examples/example-java/src/main/java/example/PresignedUploadController.java
package example;

import io.micronaut.http.HttpResponse;
import io.micronaut.http.MediaType;
import io.micronaut.http.annotation.Body;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Post;
import io.micronaut.objectstorage.ObjectStorageOperations;
import io.micronaut.objectstorage.request.CreatePresignedUploadRequest;
import io.micronaut.objectstorage.response.PresignedUpload;

import java.time.Duration;
import java.time.Instant;
import java.util.List;
import java.util.Map;

//tag::beginclass[]
@Controller("/uploads")
public class PresignedUploadController {

    private final ObjectStorageOperations<?, ?, ?> objectStorage;

    public PresignedUploadController(ObjectStorageOperations<?, ?, ?> objectStorage) {
        this.objectStorage = objectStorage;
    }
//end::beginclass[]

    //tag::presigned-upload[]
    @Post(uri = "/signed", consumes = MediaType.APPLICATION_JSON, produces = MediaType.APPLICATION_JSON)
    public HttpResponse<PresignedUploadResponse> create(@Body CreateUploadCommand command) {
        CreatePresignedUploadRequest request =
            new CreatePresignedUploadRequest("profiles/" + command.userId() + ".png", Duration.ofMinutes(5)); // <1>
        request.setContentType(MediaType.IMAGE_PNG);
        request.setMetadata(Map.of("owner", command.userId()));

        PresignedUpload signedUpload = objectStorage.createPresignedUpload(request)
            .orElseThrow(() -> new IllegalStateException("This provider does not support pre-signed uploads"));

        return HttpResponse.ok(new PresignedUploadResponse( // <2>
            signedUpload.getUri().toString(),
            signedUpload.getMethod(),
            signedUpload.getHeaders(),
            signedUpload.getExpiration()
        ));
    }
    //end::presigned-upload[]

    public record CreateUploadCommand(String userId) {
    }

    public record PresignedUploadResponse(String url,
                                          String method,
                                          Map<String, List<String>> headers,
                                          Instant expiresAt) {
    }
}
//tag::endclass[]

File: doc-examples/example-kotlin/src/main/kotlin/example/PresignedUploadController.kt
package example

import io.micronaut.http.HttpResponse
import io.micronaut.http.MediaType
import io.micronaut.http.annotation.Body
import io.micronaut.http.annotation.Controller
import io.micronaut.http.annotation.Post
import io.micronaut.objectstorage.ObjectStorageOperations
import io.micronaut.objectstorage.request.CreatePresignedUploadRequest
import java.time.Duration
import java.time.Instant

//tag::beginclass[]
@Controller("/uploads")
open class PresignedUploadController(private val objectStorage: ObjectStorageOperations<*, *, *>) {
//end::beginclass[]

    //tag::presigned-upload[]
    @Post(uri = "/signed", consumes = [MediaType.APPLICATION_JSON], produces = [MediaType.APPLICATION_JSON])
    open fun create(@Body command: CreateUploadCommand): HttpResponse<PresignedUploadResponse> {
        val request = CreatePresignedUploadRequest("profiles/${command.userId}.png", Duration.ofMinutes(5)) // <1>
        request.setContentType(MediaType.IMAGE_PNG)
        request.setMetadata(mapOf("owner" to command.userId))

        val signedUpload = objectStorage.createPresignedUpload(request)
            .orElseThrow { IllegalStateException("This provider does not support pre-signed uploads") }

        return HttpResponse.ok(
            PresignedUploadResponse( // <2>
                signedUpload.uri.toString(),
                signedUpload.method,
                signedUpload.headers,
                signedUpload.expiration
            )
        )
    }
    //end::presigned-upload[]

    data class CreateUploadCommand(val userId: String)

    data class PresignedUploadResponse(
        val url: String,
        val method: String,
        val headers: Map<String, List<String>>,
        val expiresAt: Instant
    )
}
//tag::endclass[]
